MailSendFilterPlugin.php

```diff
@@ -0,0 +1,312 @@
+<?php
+
+/**
+ * @file MailSendFilterPlugin.php
+ *
+ * Copyright (c) 2024 Simon Fraser University
+ * Copyright (c) 2024 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class MailSendFilterPlugin
+ *
+ * @brief Main plugin class, setups the email override and settings.
+ */
+
+namespace APP\plugins\generic\mailSendFilter;
+
+use APP\core\Application;
+use APP\notification\NotificationManager;
+use APP\plugins\generic\mailSendFilter\classes\MailFilter;
+use APP\plugins\generic\mailSendFilter\classes\SettingsForm;
+use Illuminate\Database\Query\Builder;
+use Illuminate\Mail\SentMessage;
+use Illuminate\Support\Collection;
+use Illuminate\Support\Facades\DB;
+use Illuminate\Support\Facades\Mail;
+use PKP\core\JSONMessage;
+use PKP\linkAction\LinkAction;
+use PKP\linkAction\request\AjaxModal;
+use PKP\linkAction\request\RedirectAction;
+use PKP\mail\Mailable;
+use PKP\plugins\GenericPlugin;
+use ReflectionClass;
+use SplFileObject;
+
+class MailSendFilterPlugin extends GenericPlugin
+{
+    // Fake ID for the threshold that deals with users with no roles
+    public const THRESHOLD_UNASSIGNED_ROLE = -1;
+    // Fake ID for the threshold that deals with users who are assigned to at least one submission
+    public const THRESHOLD_ASSIGNED_SUBMISSION = -2;
+    /** @var array<string,string> Description map of custom thresholds */
+    public array $customThresholds = [
+        self::THRESHOLD_UNASSIGNED_ROLE => 'user.role.none',
+        self::THRESHOLD_ASSIGNED_SUBMISSION => 'user.with.submission'
+    ];
+    /** @var string[] List of email keys which won't be filtered by the plugin */
+    private array $passthroughMailKeys = [];
+
+    /**
+     * @copydoc Plugin::register
+     *
+     * @param null|int $mainContextId
+     */
+    public function register($category, $path, $mainContextId = null): bool
+    {
+        $success = parent::register($category, $path, $mainContextId);
+        if (!$success || !$this->getEnabled()) {
+            return $success;
+        }
+
+        $this->setupMailOverride();
+        $this->passthroughMailKeys = json_decode($this->getSetting($this->getCurrentContextId(), 'passthroughMailKeys')) ?: [];
```

Comment:
I think `json_decode` doesn't expect a potential `null` as input, so this might throw a warning (if not now then in future) if the settings form hasn't been saved yet.

Reply:
The plugin has default settings, so a null value isn't going to appear here.